File: backend/app/schemas/universal_tracking.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict, Field

# -----------------
# Batch Master
# -----------------
class UniversalBatchMasterBase(BaseModel):
    item_id: uuid.UUID
    batch_number: str = Field(..., max_length=128)
    mfg_date: date | None = None
    expiry_date: date | None = None
    shelf_life_days: int | None = None
    status: str = Field("active", max_length=32)
    cost_multiplier: float = Field(1.0, ge=0)

class UniversalBatchMasterCreate(UniversalBatchMasterBase):
    pass

class UniversalBatchMasterResponse(UniversalBatchMasterBase):
    id: uuid.UUID
    tenant_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

# -----------------
# Serial Master
# -----------------
class UniversalSerialMasterBase(BaseModel):
    item_id: uuid.UUID
    batch_id: uuid.UUID | None = None
    serial_number: str = Field(..., max_length=128)
    status: str = Field("available", max_length=32)
    warehouse_id: uuid.UUID | None = None
    bin_id: uuid.UUID | None = None
    warranty_expiry: date | None = None

class UniversalSerialMasterCreate(UniversalSerialMasterBase):
    pass

class UniversalSerialMasterResponse(UniversalSerialMasterBase):
    id: uuid.UUID
    tenant_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/universal_tracking.ts
export interface UniversalBatchMaster {
    id: string;
    tenant_id: string;
    item_id: string;
    batch_number: string;
    mfg_date?: string;
    expiry_date?: string;
    shelf_life_days?: number;
    status: string;
    cost_multiplier: number;
    created_at: string;
    updated_at: string;
}

export interface UniversalSerialMaster {
    id: string;
    tenant_id: string;
    item_id: string;
    batch_id?: string;
    serial_number: string;
    status: string;
    warehouse_id?: string;
    bin_id?: string;
    warranty_expiry?: string;
    created_at: string;
    updated_at: string;
}
